Expand moose code standard (#32132)
idaholab/moose: framework/doc/content/sqa/framework_scs.md

    doSomething();
```

## Const-correctness

We expect contributions to follow strict const-correctness.

- Use const to accurately express intent, invariants, and ownership:
- Mark function parameters const whenever they are not modified.
- Mark member functions const whenever they do not modify observable object state.
- Prefer returning const references when mutation is not intended.
- Avoid casting away const (e.g., `const_cast`) except in narrowly justified, well-documented cases.

Const-correctness is not cosmetic. It:

- Improves API clarity and self-documentation
- Enables safer refactoring and compiler diagnostics
- Allows the compiler to reason more effectively about aliasing and optimization
- Prevents accidental state mutation, especially in generic and templated code
- Code that is logically const but not marked const will generally be treated as a design issue during review.

> 

Reviewer: ```suggestion

Code that is logically const but not marked const will generally be treated as a design issue during review.
```

> ## Index Variables in Looping Constructs

Reviewer 2: ```suggestion

## Index Variables in Looping Constructs
```

```C++
  // Looping to a scalar index
  for (MooseIndex(n_nodes) i = 0; i < nodes; ++i)
    // ... Do something with index i

  // Looping to a container size (where range-for isn't sufficient)
  for (MooseIndex(vector) i = 0; i < vector.size(); ++i)
    // ... Do something with index i
```

In addition, a few convenience methods are available to help generate integer ranges:

```C++
#include "libmesh/int_range.h"

> // Looping through n1, n1+1, n1+2, ..., n2-1, n2

Reviewer 3: ```suggestion
// Looping through n1, n1+1, n1+2, ..., n2-1
```

for (auto i : make_range(n1, n2))
  // ... Do something with index i

// Looping through 0, 1, 2, ..., n
for (auto i : make_range(n))
  // ... Do something with index i

> // Looping through 0, 1, 2, ..., v.size()

Reviewer 3: ```suggestion
// Looping through 0, 1, 2, ..., v.size()-1
```

for (const auto i : index_range(v))
  // ... Do something with index i
```

## Structured binding

Prefer structured binding when unpacking aggregate types and tuple-like objects.

An example for aggregate type structured binding:

```C++
struct S
{
  int x;
  double y;
};

S s{1, 2.0};

// Bind by copying
auto [a, b] = s;

// Bind by reference
const auto & [a_ref, b_ref] = s;
```

An example for tuple-like object binding,

```C++
std::tuple<int, Real, Point> t = {1, 0.5, Point(0.1, 0.2, 0.3)};
const auto & [a, b, c] = t;
```

Tuple-like object includes `std::pair`, `std::tuple`, `std::array` etc.

Using structured binding for unpacking sub-objects in a loop makes the intention very clear, i.e.,

```C++
std::vector<std::tuple<dof_id_type, unsigned int, BoundaryID>> my_vec;
for (const auto & [elem_id, side, bnd] : my_vec)
  // ...
```

## Aliasing

Consider using `using` (over `typedef`) to define type aliases to improve code readability:

```C++
std::vector<std::tuple<dof_id_type, unsigned int, BoundaryID>> my_vec;

// Simplify code and clarify intention by defining an alias
using ElemSide = std::tuple<dof_id_type, unsigned int, BoundaryID>;
std::vector<ElemSide> my_vec2;
```

> Remember `using` also works with template.

Reviewer 2: ```suggestion
Remember that `using` also works with template.
```

```C++
#include "libmesh/int_range.h"

// Looping through n1, n1+1, n1+2, ..., n2-1
for (const auto i : make_range(n1, n2))
  // ... Do something with index i

// Looping through 0, 1, 2, ..., n-1
for (const auto i : make_range(n))
  // ... Do something with index i

// Looping through 0, 1, 2, ..., v.size()-1
for (const auto i : index_range(v))
  // ... Do something with index i
```

## Structured binding

Prefer structured binding when unpacking aggregate types and tuple-like objects.

An example for aggregate type structured binding:

```C++
struct S
{
  int x;
  double y;
};

S s{1, 2.0};

// Bind by copying
auto [a, b] = s;

// Bind by reference
const auto & [a_ref, b_ref] = s;
```

An example for tuple-like object binding,

```C++
std::tuple<int, Real, Point> t = {1, 0.5, Point(0.1, 0.2, 0.3)};
const auto & [a, b, c] = t;
```

Tuple-like object includes `std::pair`, `std::tuple`, `std::array` etc.

Using structured binding for unpacking sub-objects in a loop makes the intention very clear, i.e.,

```C++
std::vector<std::tuple<dof_id_type, unsigned int, BoundaryID>> my_vec;
for (const auto & [elem_id, side, bnd] : my_vec)
  // ...

> ```

Author: I did some search on this. In C++17, `_` is just a normal identifier instead of a placeholder. The compilers we currently support probably don’t warn per-binding if any binding is used. For that reason, I guess I will not mention the use `_` in the guidelines for now.

#include <system_library>
```

## Free function vs. member function vs. static member function

When adding a new function, carefully decide whether it should be

- a **non-static member function** (operates on an object),
- a **static member function** (conceptually belongs to a type, but not to an instance), or
- a **free function** (algorithm or operation external to any one type).

!alert note title=Core rule
Make it a non-static member function only if it fundamentally depends on the object's invariants or representation.

### Use a **non-static member function** when the operation:

- Maintains or relies on class invariants

  ``` cpp
  class Matrix
  {
  public:
    void resize(int m, int n); // must keep internal consistency
  };
  ```

>   If the function mutates internal state, must preserve invariants, or requires intimate knowledge of representation, it should be a non-static member.

Reviewer 4: get rid of invariants. Alternatively instead of replacing invariants everywhere you could give a very clear definition of invariants at the top of this section

  ``` cpp
  class Tensor
  {
  public:
    static Tensor identity(unsigned int dim);
    static Tensor fromEulerAngles(double a, double b, double c);
  };
  ```

  These are type-level operations: factories, validators, canonical constructors, or utilities closely tied to the abstraction.

- Provides behavior naturally namespaced under the class

  ``` cpp
  Elem::build(type, dim);
  FEProblemBase::validParams();
  ```

  Static members improve discoverability, communicate conceptual ownership, and avoid polluting surrounding namespaces, even though no object data is used.

- Logically groups algorithms that are tightly coupled to a type's meaning

  Especially in large frameworks like MOOSE, static members can act as a *semantic namespace* that documents intent better than a free function.

  Think of static members as answering:

  > "Does this function describe the concept of this class, even though it doesn't operate on an instance?"

  If yes, a static member is often the right choice.

### Use a **free function** when the operation is:

- Symmetric

  ``` cpp
  dot(a, b);
  distance(p, q);
  misorientation(g1, g2);
  ```

>   If no operand is conceptually dominant, avoid `a.dot(b)`.

Reviewer 4: If no operand is conceptually dominant, e.g. avoid `a.dot(b)`.

- An algorithm over data

  ``` cpp
  norm(v);
  det(A);
  project(u, v);
  ```

>   If it takes objects as inputs, produces a value, and does not manage invariants, prefer a free function (STL style).

Reviewer 4: ~invariants~

- Meant to be extensible across unrelated types

  ``` cpp
  template <class T>
  auto norm(const T & x);
  ```

  Free functions enable generic programming, ADL customization, and retroactive extension to foreign types.

- Conceptually "between" types

  ``` cpp
  intersect(mesh, plane);
  assemble(system, element);
  ```

  These don't belong to either type alone.

### Decision checklist

> 1.  Does it mutate the object or enforce invariants? Yes → non-static member

Reviewer 4: ~invariants~

3. Is the operation symmetric or algorithmic? Yes → free function
4. Do I want/expect others to extend this behavior for new types? Yes → free function
5. If you’re unsure → default to free function.

As a reminder, helper free functions defined only in the translation unit (i.e., `.C` files) should be marked as `static` or encapsulated in an anonymous namespace to force static linking.

## Access control: public, protected, and private

Choose the narrowest access level that correctly expresses intent.

Public members and methods form the stable API of a class. They define what users are allowed to rely on. Use `public` only for:

- Operations that are safe, well-defined, and intended for external use
- Methods that preserve class invariants
- Behavior you are willing to support long-term

Public interfaces should be minimal, const-correct, and clearly documented. Adding something to `public` is a commitment.

Prefer `private` by default. Use `private` for:

- Representation details and internal state
- Helper functions and implementation mechanics
- Anything that should not be relied upon by derived classes or external code

> Private members allow internal refactoring without breaking users and are the strongest tool for enforcing invariants.

Author: ```suggestion
Private members allow internal refactoring without breaking user assumptions and are the strongest tool for enforcing invariants.
```

`protected` exists to support controlled customization via inheritance, not general access. Use `protected` only when:

- A derived class must override or extend behavior
- The base class explicitly documents a customization or extension point
- The invariant contract between base and derived classes is clear
- Avoid exposing data members as `protected`. Prefer `protected` virtual functions or `protected` helper methods instead.

If a member is not intended to be used or overridden by derived classes, it should **not** be `protected`.

In summary:

- Default to `private`
- Promote to `public` only when there is a clear, justified use case

> - Treat `protected` as part of the class’s internal design contract, not its public API

Reviewer: Most people, including me, don't know what "internal design contract" means

Author: Yeah, I rephrased it.

- Prefer composition over inheritance when possible

Code reviews may request access-level reductions if members are more visible than necessary.

## Generic Programming and Compile-Time Constants

Prefer generic programming and compile-time constants when they improve correctness, expressiveness, or performance. Use templates to express behavior that is truly type- or dimension-independent.

Prefer:

- Function and class templates for algorithms that operate uniformly across types
- `auto` and template parameter deduction to reduce redundancy
- Concepts (not available yet in C++17) or `static_assert` to document and enforce template requirements

Avoid:

- Templates used only to avoid writing a small amount of code
- Deep or obscure template metaprogramming that obscures intent
- Encoding runtime variability as template parameters without a clear benefit

Templates should make code more general and clearer, not harder to understand.

Prefer compile-time constants when values are known at compile time and semantically fixed.

Use:

> - constexpr or consteval for values and functions known at compile time

Reviewer 3: ```suggestion
- `constexpr` or `consteval` for values and functions known at compile time
```

  class Matrix
  {
  public:
    void resize(int m, int n); // must keep internal consistency
  };
  ```

  If the function mutates internal state, must preserve class invariants, or requires intimate knowledge of representation, make it a member function.

- Is conceptually “owned” by the type

  ```C++
  particle.update(dt);
  tensor.transpose();
  ```

  Ask: “Would users naturally discover this by typing obj.?” If so, make it a member function.

- Needs privileged access: If the function would otherwise require: friendship, access to private data, or bypassing encapsulation, make it a member (or a carefully chosen friend).

Use a free function for

- Symmetric operations

  ```C++
  dot(a, b);
  distance(p, q);
  misorientation(g1, g2);
  ```
  If no operand is conceptually dominant: Avoid `a.dot(b)` and prefer `dot(a, b)`.

- Algorithms over data

  ```C++
  norm(v);
  det(A);
  project(u, v);
  ```

  If it takes an object as input, produces a value, and does not modify the object, make it a free function. This aligns with the STL design: containers own storage; algorithms are free.

- Extensibility across types

  ```C++
  template<class T>
  auto norm(const T & x);
  ```

>   Free functions allow generic programming, ADL-based customization, extension to foreign types. Member functions cannot be added retroactively.

Author: ```suggestion
  Free functions allow generic programming, customization via argument-dependent lookup, extension to foreign types. Member functions cannot be added retroactively.
```

Prefer `private` by default. Use `private` for:

- Representation details and internal state
- Helper functions and implementation mechanics
- Anything that should not be relied upon by derived classes or external code

Private members allow internal refactoring without breaking user assumptions and are the strongest tool for enforcing invariants.

`protected` exists to support controlled customization via inheritance, not general access. Use `protected` only when:

- A derived class must override or extend behavior
- The base class explicitly documents a customization or extension point
- The invariant contract between base and derived classes is clear

Avoid exposing data members as `protected`. Prefer `protected` virtual functions or `protected` helper methods instead.

If a member is not intended to be used or overridden by derived classes, it should **not** be `protected`.

In summary:

- Default to `private`
- Promote to `public` only when there is a clear, justified use case
- Treat `protected` as part of the class’s internal design contract, not its public API

> - Prefer composition over inheritance when possible

Author: ```suggestion
- Consider composition as an alternative to inheritance, especially when behavior can be delegated rather than extended.
```
